pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/FunctionMetaDataManager.java

```diff
@@ -206,14 +210,29 @@ public synchronized void updateFunctionOnLeader(FunctionMetaData functionMetaDat
         } else {
             needsScheduling = processUpdate(functionMetaData);
         }
-        Request.ServiceRequest serviceRequest = Request.ServiceRequest.newBuilder()
-                .setServiceRequestType(delete ? Request.ServiceRequest.ServiceRequestType.DELETE : Request.ServiceRequest.ServiceRequestType.UPDATE)
-                .setFunctionMetaData(functionMetaData)
-                .setWorkerId(workerConfig.getWorkerId())
-                .setRequestId(UUID.randomUUID().toString())
-                .build();
+        String key = FunctionCommon.getFullyQualifiedName(functionMetaData.getFunctionDetails());
+        byte[] toWrite;
+        if (workerConfig.getUseCompactedMetadataTopic()) {
+            if (delete) {
+                toWrite = "".getBytes();
+            } else {
+                toWrite = functionMetaData.toByteArray();
+            }
+        } else {
+            Request.ServiceRequest serviceRequest = Request.ServiceRequest.newBuilder()
+                    .setServiceRequestType(delete ? Request.ServiceRequest.ServiceRequestType.DELETE : Request.ServiceRequest.ServiceRequestType.UPDATE)
+                    .setFunctionMetaData(functionMetaData)
+                    .setWorkerId(workerConfig.getWorkerId())
+                    .setRequestId(UUID.randomUUID().toString())
+                    .build();
+            toWrite = serviceRequest.toByteArray();
+        }
         try {
-            lastMessageSeen = exclusiveLeaderProducer.send(serviceRequest.toByteArray());
+            lastMessageSeen = exclusiveLeaderProducer.newMessage()
+                    .key(key)
```

Comment:
This will add a key for both the old format and new format right? Not sure we should add a key for the old format.  Perhaps, whether checking whether the message has a key or not can be used to determine whether its the old format or new format.

Reply:
Is that really useful?

Reply:
Changed